text/0003-templates.md

```diff
@@ -0,0 +1,447 @@
+
+```
+node_modules/.pnpm/@example+patches/node_modules/@example/patches/patches/express@4.18.1.patch
+```
+
+### Combining Templates
+
+Templates may not be templated from other templates. Instead of an inheritance hierarchy, a package may refer to multiple templates with later entries in the list taking precedence.
+
+For example, if both `@organization/authoring-metadata` and `@team/authoring-metadata` have an `author` field, the value from `@team/authoring-metadata` will be used.
+
+```json5
+{
+  "name": "@example/simple",
+  "pnpm": {
+    "templates": {
+      "authoring": [
+        "@organization/authoring-metadata@0.1.0",
+        "@team/authoring-metadata@0.1.0"
+      ],
+    }
+  }
+}
+```
+
+Avoiding inheritance simplifies implementation and usability in several ways.
+
+- Suppose a new version of pnpm adds `pnpm.templates` config options. In an inheritance model, using these new options on a template would force consumers to a greater minimum version of pnpm.
+- Complex circular resolution during installation is avoided.
+- Template loading is more performant. The requirement to declare all required templates up front enables fetching in a single parallelized step, rather than a waterfall fetch at each level of a theoretical template inheritance hierarchy.
+- It becomes possible to reference multiple templates the package's author does not control with less ambiguity. In an inheritance-based mechanism, multiple inheritance would be necessary for such functionality. However, multiple inheritance can lead to [diamond-shaped problems](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem) and less clarity as to what the final value of the field should be.
+
+A few of the problems above can be mitigated by rendering the template package itself before publishing, but this introduces other problems. Toolkit templates install dependencies in the referencing package as `devDependencies`, which isn't desirable in other templates.
+
+The existing design is inspired by the concept of ["_composition over inheritance_"](https://en.wikipedia.org/wiki/Composition_over_inheritance), which provides reusability without defining difficult to change relationships between packages.
+
+### Specifiers
+
+Templates can be specified through:
+
+1. The `name@version` syntax to refer to an external template fetched from a registry.
+2. The `name@workspace:` syntax to refer to a template that's also a workspace package.
+3. The `file:<path>` syntax to refer to an on-disk path.
+
+### Shared Workspace Configuration
```

Comment:
can we make this a separate RFC? As far as I understand all the other parts were discussed an agreed upon.

Reply:
> can we make this a separate RFC?

Can do! Done in 9cd64ecafb74ccecef010079e0c550d7f8bc062c. I forgot to mention that this part was in response to https://github.com/pnpm/rfcs/pull/3#issuecomment-1457373197, if that helps.

> As far as I understand all the other parts were discussed an agreed upon.

That mostly sounds right. We discussed the potential need to serialize templates to the lockfile, but there's some new details around the integrity hash that might be contentious. https://github.com/pnpm/rfcs/pull/3/files#r1241478370